### PR TITLE
Add working spack build with build cache.

### LINKED ISSUE #601: Spack CI failures
One of the Spack-based CIs started to fail recently. Namely, `hiop@develop~mpi+raja~shared~kron~sparse`

This happens for all the PRs/branches. It reports 
```
/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/cmake-3.23.3-
          u7spcudumu4d6olw2rjdhugl6g6z6ndq/bin/cmake: error while loading share
          d libraries: libxml2.so.2: cannot open shared object file: No such fi
          le or directory
```

### PR DESCRIPTION
**Pull request by cameronrutherford**
Closes #601

### CODE REVIEW

**Comment by cnpetra** [comment]
FYI, one of the spack ubuntu builds fails.

**Comment by nychiang** [comment]
@cameronrutherford CURL cannot be found from the path [(see here)](https://github.com/LLNL/hiop/actions/runs/10690678532/job/29635606533?pr=695#step:10:172), but I see spack has specified to install curl.

**Comment by cameronrutherford** [comment]
> @cameronrutherford CURL cannot be found from the path [(see here)](https://github.com/LLNL/hiop/actions/runs/10690678532/job/29635606533?pr=695#step:10:172), but I see spack has specified to install curl.

Yes... not really sure what's going on but I haven't had time to try and debug. I'm thinking if we let spack build curl this will resolve as the bug is with using external curl. 

Just need to exclude it the same way we exclude finding Python..

**Comment by cnpetra** [comment]
it loads curl and openssl, but now fails with 
``CMake Error: The source directory "/__w/hiop/hiop" does not appear  to contain CMakeLists.txt. ``

**Comment by cameronrutherford** [comment]
https://github.com/spack/spack/pull/46378 - I thought this would fix the issue. I think it's a reasonable fix to the spack package, but didn't resolve the issue.

What's frustrating here is that the error log isn't printing it's full output. I can't even see the full CMake line. I went down this path because there was an extra `''` in the CMake line, but now I can't even tell what command is being invoked.

GitHub actions further complicates the problem because it's all running inside a VM, so unless I spend money on Azure it's hard to reproduce.

I will take a look again later with a fresh perspective and see what I can figure out. Confusing why we are only seeing this sort of issue with HiOp, not ExaGO or other packages...

**Comment by cnpetra** [comment]
This time fails with 
``CMake Error: The source directory "/home/runner/work/hiop/hiop" does not appear to contain CMakeLists.txt.``
Looks like it is a path concatenation issue.

If it is too cumbersome to make this work, then I suggest to disable this build. All the functionality is tested by the LLNL CI.

**Comment by cameronrutherford** [comment]
> This time fails with `CMake Error: The source directory "/home/runner/work/hiop/hiop" does not appear to contain CMakeLists.txt.` Looks like it is a path concatenation issue.
> 
> If it is too cumbersome to make this work, then I suggest to disable this build. All the functionality is tested by the LLNL CI.

I can't believe I missed it, but we simply weren't cloning the repo in the spack pipelines. The pipelines are really flying here and complete in ~5 minutes

**Comment by cameronrutherford** [comment]
MPI CPU builds are running, passing tests and being pushed to a buildcahe for each branch in your repo. All in under 5 minutes if we don't have to re-build dependencies!

File: .github/workflows/spack_build.yml [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+          # Once we move submodule deps into spack, we can do some more builds
+          # Also need to change build script to use spack from base image
+          submodules: true
+
+      - name: Clone Spack
+        run: |
+          # TODO: Move back to spack@develop once patch in in develop
+          git clone https://github.com/cameronrutherford/spack.git
+          pushd spack
+          git checkout 838f2b6708a28cae7b1c1de1f1e1f29a1bf2ee0d
```
Comment (cameronrutherford) [inline .github/workflows/spack_build.yml]:
```suggestion
          git clone https://github.com/spack/spack.git
```

I am pretty sure I never needed to make my spack change...

File: .github/workflows/spack_build.yml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      - name: Check for existing base images
+        run: |
+          set -e
+          CONTAINER_TAG=${{ env.BASE_VERSION }}
+          OCI_URL="${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.BASE_VERSION }}"
+          echo Checking for CONTAINER_TAG $CONTAINER_TAG
+          skopeo inspect \
+              docker://$OCI_URL \
+              --raw \
+              --creds "${{ env.USERNAME }}:${{ secrets.GITHUB_TOKEN }}" \
```
Comment (nychiang) [inline .github/workflows/spack_build.yml:43]: where do we provide the token? 
I'd like to learn more about how you create the user `hiop-bot` and the dependencies in the container, but we can do it offline.

**Review by nychiang — APPROVED** [review]
LGTM. It turns out I need to update my knowledge from Cameron.

File: .github/workflows/spack_build.yml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  TMPDIR: /opt/spack-cache
+  # Our repo name contains upper case characters, so we can't use ${{ github.repository }}
+  IMAGE_NAME: hiop
+  USERNAME: hiop-bot
+  BASE_VERSION: ubuntu-24.04-fortran
+
+jobs:
+  base_image_build:
+    runs-on: ubuntu-24.04
+    permissions:
```
Comment (cameronrutherford) [inline .github/workflows/spack_build.yml:21]: @nychiang tokens are automatically created by pipelines with permissions based on this YAML config.

File: .github/workflows/spack_build.yml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+env:
+  SPACK_COLOR: always
+  REGISTRY: ghcr.io/llnl
+  SPACK_CACHE: /opt/spack-cache
+  tempdir: /opt/spack-cache
+  TMP: /opt/spack-cache
+  TMPDIR: /opt/spack-cache
+  # Our repo name contains upper case characters, so we can't use ${{ github.repository }}
+  IMAGE_NAME: hiop
+  USERNAME: hiop-bot
```
Comment (cameronrutherford) [inline .github/workflows/spack_build.yml:15]: @nychiang name is arbitrary. You can pick a different name 😁

File: .github/workflows/spack_build.yml [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+          RUN apt-get update && \
+              apt-get install -y --no-install-recommends \
+                software-properties-common \
+                gpg-agent \
+                openssh-client \
+                openssh-server \
+              && rm -rf /var/lib/apt/lists/*
+          RUN add-apt-repository ppa:ubuntu-toolchain-r/test && \
+              apt-get install -y --no-install-recommends \
+                gfortran \
```
Comment (cameronrutherford) [inline .github/workflows/spack_build.yml:62]: @nychiang gfortran was needed to launch MPI based applications when I was doing this for ExaGO.

I admittedly didn't pull/test the HiOp images, but this is covered in the tutorial I followed https://github.com/spack/spack-tutorial/blob/main/tutorial_binary_cache.rst#creating-runnable-container-images

Comment (cameronrutherford) [reply .github/workflows/spack_build.yml:62]: Figuring out the dependencies was admittedly trial and error.

For ExaGo, I was running MPI enabled OPFLOW to get our `--version` output, and logically added to the base image from the error messages I got.

**Review by cameronrutherford — COMMENTED** [review]
Hopefully this clarifies your question @nychiang

**Pull request merged**